storage/bootstrap.go

```diff
@@ -86,43 +86,85 @@ func newBootstrapManager(
 	}
 }
 
+func (m *bootstrapManager) IsBootstrapping() bool {
+	m.RLock()
+	state := m.state
+	m.RUnlock()
+	return state == bootstrapping
+}
+
 func (m *bootstrapManager) IsBootstrapped() bool {
 	m.RLock()
 	state := m.state
 	m.RUnlock()
 	return state == bootstrapped
 }
 
-func (m *bootstrapManager) targetRanges(writeStart time.Time) xtime.Ranges {
+func (m *bootstrapManager) targetRanges() xtime.Ranges {
+	now := m.nowFn()
 	ropts := m.opts.RetentionOptions()
-	start := writeStart.Add(-ropts.RetentionPeriod())
-	midPoint := writeStart.
+	start := now.Add(-ropts.RetentionPeriod())
+	midPoint := now.
 		Add(-ropts.BlockSize()).
 		Add(-ropts.BufferPast()).
 		Truncate(ropts.BlockSize())
-	cutover := writeStart.Add(ropts.BufferFuture())
+	cutover := now.Add(ropts.BufferFuture())
 
 	return xtime.NewRanges().
 		AddRange(xtime.Range{Start: start, End: midPoint}).
 		AddRange(xtime.Range{Start: midPoint, End: cutover})
 }
 
-// NB(xichen): Bootstrap must be called after the server starts accepting writes.
 func (m *bootstrapManager) Bootstrap() error {
-	writeStart := m.nowFn()
-	targetRanges := m.targetRanges(writeStart)
-
+	var enqueued bool
 	m.Lock()
-	if m.state == bootstrapped {
-		m.Unlock()
-		return nil
+	switch m.state {
+	case bootstrapping:
+		// NB(r): Already bootstrapping, now a consequent bootstrap
+		// request comes in - we queue this up to bootstrap again
+		// once the current bootstrap has completed.
+		// This is an edge case that can occur if during either an
+		// initial bootstrap or a resharding bootstrap if a new
+		// reshard occurs and we need to bootstrap more shards.
+		m.pending = true
```

Comment:
Wonder if we can shortcut the current bootstrap and exit early since we need to rebootstrap anyway